### ISSUE TITLE
verifychecksum.bin cannot deal with multi-segment files

### ISSUE THREAD

**Issue by mbanck** [uptimejp/postgres-toolkit#45]
If a heap file is larger than 1GB, Postgres splits it up into several segment files, with `.1`, `.2` at the end and so on.  The block number is not reset to 0 however, but is sequential, which means that `verifychecksum.bin` fails for any multi-segment files as the block number is part of the checksum.

**Comment by mbanck** [comment]
Patch for this:

```
--- verifychecksum.c.orig       2018-01-23 14:12:23.000000000 +0100
+++ verifychecksum.c.new        2018-01-23 14:15:39.000000000 +0100
@@ -5,6 +5,7 @@
  */
 #include <stdio.h>
 #include <unistd.h>
+#include <libgen.h>
 
 #include "c.h"
 #include "pg_config.h"
@@ -37,6 +38,8 @@
   char page[BLCKSZ];
   BlockNumber blkno = 0;
   BlockNumber corrupted = 0;
+  char* filename;
+  int segno = 0;
 
   fd = open(filepath, O_RDONLY);
   if (fd <= 0)
@@ -45,9 +48,15 @@
       exit(2);
     }
 
+  filename = strdup(filepath);
+  if (strstr(basename(filename), "."))
+   {
+     segno = atoi(strstr(basename(filename), ".")+1);
+   }
+
   while (read(fd, page, BLCKSZ) == BLCKSZ)
     {
-      if (!verify_page(page, blkno, filepath))
+      if (!verify_page(page, segno * RELSEG_SIZE + blkno, filepath))
        {
          corrupted++;
        }
```

**Comment by snaga** [comment]
Do you need this fix in verifychecksum.c?

We already can handle multi-segment file with pt-verify-checksum, a wrapper for the verifychecksum command.
https://github.com/uptimejp/postgres-toolkit/blob/develop/bin/pt-verify-checksum

I would like to keep verifychecksum.c simple as much as possible.

**Comment by mbanck** [comment]
Well, I would like to have it in `verifychecksum.c`, this way you can e.g. plug it into a tar pipe with e.g. ` --to-command='verifychecksum.bin /dev/stdin > /dev/null || echo
"corrupt: $TAR_FILENAME"'`.

That said, how does `pt-verify-checksum`  handle multiple-segment files? Note that I am not talking about the `_vm` etc. files, but the `.1`, `.2` continuation files after 1GB. `pt-verify-checksum` would need to tell `verifychecksum.c` which segment it is, because the first block in `.1` has `blkno` `131073` and not `1` and as the block number is part of the checksum, you get slightly wrong ones for every block in those files. 

I found this bug while trying (I think) both on a `$PGDATA` with files more than 1GB, so I'd be surprised if it worked.

You can reproduce it with e.g. `pgench -i -s 100` for example, this should create a table > 1GB.

**Comment by mbanck** [comment]
I just tested it:

```
> initdb -k --pgdata=data 1> /dev/null 2>&1
> pg_ctl --pgdata=data start 1> /dev/null
> pgbench -i -s 80 2> /dev/null
> pt-verify-checksum -r data 2>&1 | tail -3
[2018-01-27 12:06:38] INFO: data/base/12351/16396.1: blkno 75, expected d57, found d59
[2018-01-27 12:06:38] INFO: 76 blocks corrupted in data/base/12351/16396.1.
[2018-01-27 12:06:39] INFO: Verified 943 files. 1 files corrupted. 49 files skipped.
```

I had issues with higher scale factors cause the `.1` file was too big then and the amount of errors reported by `verifychecksum.bin` seems to overload the pipe and make it and `pt-verify-checksum` hang.

**Comment by snaga** [comment]
Hi @mbanck 

I finally got what you concerned about.
You mean the segment number should be included in calculating checksums as offset of block number, however, verfiychecksum does not consider that.

I'm working on it.

**Comment by snaga** [comment]
Committed. 2b17f6067341c53f5b6f1cc04a4bbb44e3d07ced

**Comment by snaga** [comment]
Merged #53.

**Comment by mbanck** [comment]
Thanks for fixing all the issues, will you make a new postgres-toolkit release in the near future, which includes these fixes?
